src/resolvelib/providers.py

```diff
@@ -94,6 +94,21 @@ def get_dependencies(self, candidate):
         """
         raise NotImplementedError
 
+    def match_identically(self, requirements_a, requirements_b):
+        """Whether the two given requirement sets find the same candidates.
+
+        This is used by the resolver to perform tree-pruning. If the two
+        requirement sets provide the same candidates, the resolver can avoid
+        visiting the subtree again when it's encountered, and directly mark it
+        as a dead end instead.
+
+        Both arguments are iterators yielding requirement objects. A boolean
+        should be returned to indicate whether the two sets should be treated
+        as matching.
```

Comment:
This should also mention `requirements_a` is the new state, and `requirements_b` is from a known-to-fail state. This is important since the provider can implement range merging to return True when `requirements_a` is a *subset* of `requirements_b` to prune additional subtrees.

(The method name should probably change.)

Reply:
OTOH if `requirements_a` is a superset of `requirements_b`, this should be used to exclude `requirements_b` from `requirements_a` so the resolver can avoid visiting some subtrees. Maybe the interface should be defined to perform specifier merging instead.